<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This is for a case like this: you have a fasta file for one chromosome in a species,
for example Chinook salmon.  You also have another fasta file which holds the bases
in coho salmon at the homologous positions in Chinook salmon.  (In other words, the
coho fasta is exactly the same length as the Chinook fasta.)  You want to find all the
positions where you have a different base in Chinook than you do in coho, and you want
to return those all in a nice data frame."><title>Find and return the substitutions between two parallel fasta files (each with only a single sequence) — fasta_dna_diffs • ecaRbioinf</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Find and return the substitutions between two parallel fasta files (each with only a single sequence) — fasta_dna_diffs"><meta property="og:description" content="This is for a case like this: you have a fasta file for one chromosome in a species,
for example Chinook salmon.  You also have another fasta file which holds the bases
in coho salmon at the homologous positions in Chinook salmon.  (In other words, the
coho fasta is exactly the same length as the Chinook fasta.)  You want to find all the
positions where you have a different base in Chinook than you do in coho, and you want
to return those all in a nice data frame."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ecaRbioinf</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.999</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Find and return the substitutions between two parallel fasta files (each with only a single sequence)</h1>
      
      <div class="d-none name"><code>fasta_dna_diffs.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This is for a case like this: you have a fasta file for one chromosome in a species,
for example Chinook salmon.  You also have another fasta file which holds the bases
in coho salmon at the homologous positions in Chinook salmon.  (In other words, the
coho fasta is exactly the same length as the Chinook fasta.)  You want to find all the
positions where you have a different base in Chinook than you do in coho, and you want
to return those all in a nice data frame.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fasta_dna_diffs</span><span class="op">(</span><span class="va">ref</span>, <span class="va">anc</span>, OutputBigTable <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>ref</dt>
<dd><p>path to a fasta (can be gzipped) that contains the "reference" sequence.</p></dd>


<dt>anc</dt>
<dd><p>path to a fasta (can be gzipped) which represents that homologous sequence,
which will often be the "ancstral" sequence which has been aligned to the exact positions
in the reference sequence.</p></dd>


<dt>OutputBigTable</dt>
<dd><p>if TRUE, then this will return a list, element "ret" of which is
the standard return value, and element "raw_tib" of which is the full tibble with the
full sequences.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Note that upper and lower case letters are treated as the same.  So, for example
ref = A and anc = a will not be counted as a substitution.  Output will be always
uppercase.</p>
<p>Additionally, if either sequence has an N, that will not be counted as a difference
or a substitution.  In fact, all bases that are not A,C,G, or T will be discarded
and not recorded as a difference.  (In other words, we are just going for substitutions
here).</p>
<p>Note, the chromosome name (CHROM in the output) will be taken from the reference
fasta (ref).</p>
<p>Note, if the sequence name in ref is of the form &gt;chrom:p1-p2, where p1 and p2
are integers, then the positions
of all the reported differences will be shifted to start at p1.</p>
<p>It was surprising to me that these operations could be done so quickly and easily
in R.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"ecaRbioinf"</span>, <span class="st">"extdata"</span>, <span class="st">"NC_037124.1_fragment_chinook.fna.gz"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">anc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"ecaRbioinf"</span>, <span class="st">"extdata"</span>, <span class="st">"NC_037124.1_fragment_coho.fna.gz"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fdd</span> <span class="op">&lt;-</span> <span class="fu">fasta_dna_diffs</span><span class="op">(</span><span class="va">ref</span>, <span class="va">anc</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Sequence starting position: 9661479</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Percentage substitutions: 0.0214500764427991</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Eric C. Anderson.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

