<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This is particularly useful when you mapped to an early version of a genome, and
then designed assays for SNPs at positions X, Y, and Z in a, say, 500 bp
fragment.  But now you want to know the coordinates of X, Y, and Z relative
to an updated version of the genome.  So, you map the 500 bp to the updated
version of the genome, only to find that there is a long and complicated
CIGAR string that links the two.  This parses that CIGAR string to give you
the genomic coordinates of X, Y, and Z relative to the new, updated reference
genome."><title>match locations on a short query sequence to the genomic coordinates of a reference — genome_coords_from_query_and_cigar • ecaRbioinf</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="match locations on a short query sequence to the genomic coordinates of a reference — genome_coords_from_query_and_cigar"><meta property="og:description" content="This is particularly useful when you mapped to an early version of a genome, and
then designed assays for SNPs at positions X, Y, and Z in a, say, 500 bp
fragment.  But now you want to know the coordinates of X, Y, and Z relative
to an updated version of the genome.  So, you map the 500 bp to the updated
version of the genome, only to find that there is a long and complicated
CIGAR string that links the two.  This parses that CIGAR string to give you
the genomic coordinates of X, Y, and Z relative to the new, updated reference
genome."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ecaRbioinf</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.999</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>match locations on a short query sequence to the genomic coordinates of a reference</h1>
      
      <div class="d-none name"><code>genome_coords_from_query_and_cigar.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This is particularly useful when you mapped to an early version of a genome, and
then designed assays for SNPs at positions X, Y, and Z in a, say, 500 bp
fragment.  But now you want to know the coordinates of X, Y, and Z relative
to an updated version of the genome.  So, you map the 500 bp to the updated
version of the genome, only to find that there is a long and complicated
CIGAR string that links the two.  This parses that CIGAR string to give you
the genomic coordinates of X, Y, and Z relative to the new, updated reference
genome.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">genome_coords_from_query_and_cigar</span><span class="op">(</span><span class="va">cigar</span>, <span class="va">start</span>, rel_pos <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  strand <span class="op">=</span> <span class="cn">NULL</span>, query_length <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>cigar</dt>
<dd><p>the cigar string</p></dd>


<dt>start</dt>
<dd><p>the base-1 left end mapping position of the query on the reference</p></dd>


<dt>rel_pos</dt>
<dd><p>the base-1 positions (1 is the start of the query sequence) whose
corresponding positions in the reference are desired.  If NULL, all positions in the
query are returned.  It is very important to note that the position here is
relative to the original query sequence, upon which hard-clipping has not been
done, for that reason, hard-clipping consumes this original query sequence.</p></dd>


<dt>strand</dt>
<dd><p>This should be "+" or "-".  "+" indicates an alignment to the
forward strand and it means that the rel_pos on the query is correct as it
stands.  If strand is "-" then the function takes care of that by computing
a new relative position that starts counting from the other end of the query
strand.  That is why query length is required.</p></dd>


<dt>query_length</dt>
<dd><p>The length of the original query sequence.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This is definitely not written for speed.  It would not be hard to write something
super fast in a few lines of C++, but I wanted to do something that is easy
to look at.</p>
<p>For the most part, this can be done using cumulative sums on whether or not
each different CIGAR operator consumes the query or the reference.</p>
<p>It looks like the same or a similar thing can be done in the BioConductor
package function,
GenomicAlignments::mapToAlignments, but I find that stuff pretty opaque
(so many different S4 classes....I suppose if one was going to take a quarter
to learn all about it, it would be good for doing bioinformatics.  But, for
an ecologist, it is, I think, way better to spend that quarter using how to use
the tidyverse well.)</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Eric C. Anderson.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

