<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Pass this thing a tibble of markers and it will make a plot with
the marker names in the order they are given, with lines to
the genomic coordinate relative to the reference used. Assumes that all
markers are on the same chromosome."><title>Plot genomic locations of different marker sets for comparison — marker_set_locations_plot • ecaRbioinf</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plot genomic locations of different marker sets for comparison — marker_set_locations_plot"><meta property="og:description" content="Pass this thing a tibble of markers and it will make a plot with
the marker names in the order they are given, with lines to
the genomic coordinate relative to the reference used. Assumes that all
markers are on the same chromosome."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">ecaRbioinf</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.999</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Plot genomic locations of different marker sets for comparison</h1>
      
      <div class="d-none name"><code>marker_set_locations_plot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Pass this thing a tibble of markers and it will make a plot with
the marker names in the order they are given, with lines to
the genomic coordinate relative to the reference used. Assumes that all
markers are on the same chromosome.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">marker_set_locations_plot</span><span class="op">(</span><span class="va">M</span>, xlabel <span class="op">=</span> <span class="st">"position"</span>, linehang <span class="op">=</span> <span class="fl">0.07</span>,</span>
<span>  boxhang <span class="op">=</span> <span class="op">-</span><span class="fl">0.1</span>, m_name_upnudge <span class="op">=</span> <span class="fl">0.01</span>, m_name_textsize <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  kink_rel <span class="op">=</span> <span class="fl">0.08</span>, conn_line_size <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>M</dt>
<dd><p>the tibble of markers.  Must be in a particular format: 1) First column
is numeric and gives the genomic coordinate of the variant/marker in each row.
The reference genome used determines the x-coordinate of the plot. The name of this
column must be "position". 2) the second column must be named pos_predicted and be a logical vector that
says whether the position of the marker in the first column is predicted or not.
If not, set them all to FALSE.
3) Each successive
column is used to store the markers names
in a particular marker set.  The title of the column should be the name of the marker
set.  The name of each marker in the marker set should appear in the row which is at the
appropriate location at the corresponding position in the coordinate column.  4) There must
be no other columns in the tibble.
5) The second column could be a character vector
that is the chrom-name:position for the reference used.  6) The order of the marker sets/coordinates
in columns 2 and up
determines the order in which they will be placed going from the bottom to the top of
the plot. 7) Note that the position of some markers may not be known on the
reference coordinates, typically because they are from a different assembly and they
do not align to the reference assembly.  Nonetheless, their order within the
rows of markers should be reasonably well defined.  Put them in in the correct order
and leave their position row as NA.  Note that the "outermost" markers should have
non-NA positions.</p></dd>


<dt>xlabel</dt>
<dd><p>The desired label beneath the x-axis.  Typically will refer to what the
coordinate reference is, such as, "Position on chr28 from Otsh_v1.0". If left empty,
it will default to "position"</p></dd>


<dt>linehang</dt>
<dd><p>a 2-vector (or recycled if shorter) that gives the overhang amount of the
position line to the left of the leftmost markers (element 1) and to the right of the rightmost
marker, expressed as a proportion of the total distance between leftmost and rightmost marker.</p></dd>


<dt>boxhang</dt>
<dd><p>a 2-vector giving the fraction of the distance between outermost markers that
we want the box edges to be at on the left and right.  Negative moves them in toward the center
and positive moves them out.</p></dd>


<dt>m_name_upnudge</dt>
<dd><p>absolute amount by which to nudge the marker names upward within their boxes.</p></dd>


<dt>m_name_textsize</dt>
<dd><p>size of text for the marker names.  This should be fiddled with *after*
the output size has been set, to make the marker names fit in their boxes.</p></dd>


<dt>kink_rel</dt>
<dd><p>the height of the kinks (and of the ticks on the position line) as a fraction
of the total height of all the marker boxes.</p></dd>


<dt>conn_line_size</dt>
<dd><p>the size (thickness) of the lines/kinks/ticks connecting marker boxes to their positions.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Function returns a ggplot object.  To change colors of the marker sets,
add a new discrete fill scale to it, as shown in the examples. Be sure to
add <code>na.value = "transparent"</code>  and set breaks to be the reverse of the
names of columns 2 and up (see examples).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Eric C. Anderson.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

